articles/121_realtime_proposal.md

```diff
@@ -0,0 +1,215 @@
+
+Though in this document we consider the failure mode orthogonal to latency characteristics, a hard real-time failure mode is often associated with high determinism (low jitter).
+Safety-critical systems often require a hard real-time with high determinism.
+
+### Implementation choices: computer architecture and OS
+
+Processors vary in capability, number of cores, memory size, and instruction set architectures, some which are better suited to real-time performance than others.
+The selection of computer architecture may be due to cost, size, or energy efficiency, which could set a constraint for the rest of the system.
+A single robot may contain multiple computers with a variety of processor types, from the main control board to sensors.
+
+Another major consideration when designing a real-time system is the choice of operating system.
+A true "hard" real-time operating system is fully deterministic.
+"Soft" real-time operating systems are designed to meet most deadlines deterministically, but may still have an element of nondeterminism.
+Most conventional user-facing operating systems are not designed with determinism or fairness as the first priority, often because they have to support graphical displays and interactivity.
+Here are some features critical to deterministic performance that are often exposed to the user in an RTOS, but not in a conventional OS (see Wikipedia, [Real-time operating system](https://en.wikipedia.org/wiki/Real-time_operating_system)).
+For more information about why these features matter, refer to the [introductory article on real-time systems](http://design.ros2.org/articles/realtime_background.html).
+
+- Task scheduling
+
+  - Expose thread priority and deterministic schedulers that incorporate priority
+  - Allow full thread preemption
+
+- Concurrency and synchronization primitives
+
+  - Floored semaphores (allow a higher priority thread to take control of locked semaphore)
+
+- Interrupt handling
+
+  - Disable interrupts
+  - Mask interrupts (temporarily suspend interrupts in a critical section)
+
+- Memory allocation
+
+Some real-time systems have a non real-time component, e.g. for graphical display or other user interaction.
+This would require an operating system that allows non real-time threads to be scheduled alongside real-time threads.
+Because of the extra nondeterminism introduced by this requirement, this kind of system would be better suited to a soft real-time constraint.
+
+Another important constraint is which processor architectures are supported by which operating systems, and which RTOS is an appropriate choice due to performance constraints of the architecture.
+
+## Goal for ROS 2
+
+In order to support real-time systems in ROS 2, we propose an iterative approach for identifying limitations in the API that force the code path into performance bottlenecks.
```

Comment:
Hi. Not knowing anything about real-time, I stumbled on this sentence, because earlier paragraphs keep talking about determinism only, but this one suddenly jumps to performance bottlenecks only. As in:  "real-time system require determinism. Goal for ROS2 is higher performance." Also in general, I would be curious in how to make sure that later additions / changes to the ROS2 code will not break whatever state is achieved. I.e. is it possible to write regression tests for the performance / determinism guarantees? I don't know what's technically possible there, if no solution exists for automated testing, then maybe a result could be a maintained ros package / library that can be run anytime by anyone on the given boards to check whether ros2 has changed in an undesired way. In other words, don't plan to do this just once, plan to do this on every release (that changed core code), if possible.
The other part I don't quite get is that ROS2 wraps rmw implementations, so in some sense if the rmw implementation guarantees some performance/non-determinism, then you might consider writing tests that test the ros2 glue code with only a dummy rmw, even if additional integration tests can also be valuable and detect wrong assumptions about rmw implementations.